### ISSUE TITLE
a bug in your code

### ISSUE THREAD

**Issue by youngornever** [XuezheMax/flowseq#3]
I think the function ``bwdpass'' in file https://github.com/XuezheMax/flowseq/blob/master/flownmt/flows/flow.py is wrong. The line94 and line99 should be exchanged.
by the way, I have a question. What does bwdpass mean? maping a complex distribution(e.g. latent variables Z in your paper) to a simpler prior distribution (e.g. latent variables V in Eq.5 in your paper)?
finally, thanks for your great work and code release.

**Comment by XuezheMax** [comment]
First, your understanding about bwdpass (and fwdpass) is correct. bwdpass is the mapping form complex distribution (data or latent z) to simple prior (gaussian). 
Second, the code in the function ``bwdpass'' in file "flow.py" is not a bug. Function "init" is used for initializing the parameters with a batch of data.

**Comment by youngornever** [comment]
Thanks for your reply. 
In the fwdpass function,  if inverse, we can't init the inverse flow. BUT, in the bwdpass,  I think the inverse case is forward flow,  in which we init the forward flow. 
So, in both fwdpass & bwdpass, we can only init the forward flow, and can not init the inverse flow.

**Comment by XuezheMax** [comment]
If a flow is in "inverse" mode, it indicates that its bwdpass is efficiently implemented. Otherwise, its fwdpass is efficient. For a flow with "inverse" mode, in the training case, it is supposed to perform bwdpass, including parameter initialization.

**Comment by Gandor26** [comment]
> If a flow is in "inverse" mode, it indicates that its bwdpass is efficiently implemented. Otherwise, its fwdpass is efficient. For a flow with "inverse" mode, in the training case, it is supposed to perform bwdpass, including parameter initialization.

Hello, thanks for your great work and sharing such a well-structured codebase. I am still confused about the "inverse mode". You said it indicates different methods that are implemented efficiently, but as far as I have seen in the code, forward and backward methods look quite similar. Could you elaborate on the purpose of "inverse" keyword?

**Comment by XuezheMax** [comment]
In our implementation, the forward function in a flow is always the efficient transformation, and the backward is the corresponding inverse transformation.
For some flows, the backward is also efficient (like the affine coupling layers in glow), but some are not (like autogressive flows).

There are generally two functions that a flow need to have: 1) computing the density of a data x. It requires to map x to the prior (gaussian) distribution. 2) sampling. It needs to map a gaussian noise back to data x.
Based on the two mapping directions, we define the fwdpass and bwdpass.

In practice, some flows requires efficient bwdpass because they need to estimate density frequently, like glow in the training case. In this case, we set it to the inverse mode to make the bwdpass efficient.

To sum up, defining the inverse code is to facilitate coding. When implementing a new flow, one only needs to write the forward and corresponding backward functions. Based on the usage of the flow, setting the inverse code to the desired mode.
